### ISSUE TITLE
请教，预训练阶段的具体细节

### ISSUE THREAD

**Issue by songbaiTalk** [ymcui/Chinese-LLaMA-Alpaca#9]
您在固定transformer参数，仅训练embedding的阶段的代码思路是否能提供一下？是否类似这样的做法：
import torch
import torch.optim as optim

optimizer = optim.Adam(model.embeddings.parameters(), lr=1e-3)

for epoch in range(num_epochs):
    for batch in data:
        inputs, labels = batch
        
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = loss_function(outputs, labels)
        loss.backward()
        
        optimizer.step()

**Comment by airaria** [comment]
> 您在固定transformer参数，仅训练embedding的阶段的代码思路是否能提供一下？是否类似这样的做法： import torch import torch.optim as optim
> 
> optimizer = optim.Adam(model.embeddings.parameters(), lr=1e-3)
> 
> for epoch in range(num_epochs): for batch in data: inputs, labels = batch
> 
> ```
>     optimizer.zero_grad()
>     outputs = model(inputs)
>     loss = loss_function(outputs, labels)
>     loss.backward()
>     
>     optimizer.step()
> ```

首先设置requries_grad

```python
trainable=['embed_tokens']
for name, parameter in model.named_parameters():
    if not any(p in name for p in trainable):
        parameter.requires_grad_(False)
````

之后在optimizer的初始化中大致是这样
```python
optimizer = optim.Adam([p for p in model.parameters() if p.requires_grad], lr=1e-3)
```
不过这一步是Transformers里的trainer自动完成的，所以实际上我们只要做第一步就行了

**Comment by songbaiTalk** [comment]
> > 您在固定transformer参数，仅训练embedding的阶段的代码思路是否能提供一下？是否类似这样的做法： import torch import torch.optim as optim
> > optimizer = optim.Adam(model.embeddings.parameters(), lr=1e-3)
> > for epoch in range(num_epochs): for batch in data: inputs, labels = batch
> > ```
> >     optimizer.zero_grad()
> >     outputs = model(inputs)
> >     loss = loss_function(outputs, labels)
> >     loss.backward()
> >     
> >     optimizer.step()
> > ```
> 
> 首先设置requries_grad
> 
> ```python
> trainable=['embed_tokens']
> for name, parameter in model.named_parameters():
>     if not any(p in name for p in trainable):
>         parameter.requires_grad_(False)
> ```
> 
> 之后在optimizer的初始化中大致是这样
> 
> ```python
> optimizer = optim.Adam([p for p in model.parameters() if p.requires_grad], lr=1e-3)
> ```
> 
> 不过这一步是Transformers里的trainer自动完成的，所以实际上我们只要做第一步就行了

非常感谢！
